torch/distributed/_tensor/experimental/local_map.py

```diff
@@ -0,0 +1,174 @@
+InputPlacements = Union[PlacementType, Tuple[PlacementType, ...]]
+OutputPlacements = Union[PlacementType, Tuple[PlacementType, ...]]
+
+
+def local_map(
+    func: Callable,
+    out_placements: OutputPlacements,
+    in_placements: Optional[InputPlacements] = None,
+    device_mesh: Optional[DeviceMesh] = None,
+    *,
+    redistribute_inputs: bool = False,
+):
+    """
+    ``local_map`` is an experimental API that allows users to apply on :class:`DTensors`
+    a function that is written to be applied on :class:`~torch.Tensors`.
+
+    Args:
+        func (Callable): the function to be applied on each local shard of
+            :class:`DTensor`s.
+        out_placements (Union[`PlacementType`, Tuple[`PlacementType`, ...]]):
+            the desired placements of the output :class:`DTensor`s. If the `output` of
+            `func` is a Python collection, the `out_placements` will be a Tuple of
+            `PlacementType` values 1:1 mapping to the flattened `output`. For
+            :class:`Tensor` output, the corresponding `PlacementType` will be its
+            placements (a `Tuple[Placement]` value). For non-:class:`Tensor` output,
+            the `PlacementType` will be `None`.
+        in_placements (Union[`PlacementType`, Tuple[`PlacementType`, ...]], optional):
+            the required placements of the input :class:`DTensor`s. If not specified,
+            the input :class:`DTensor` will not be redistributed before passing its local
+            tensor to `func`. Similarly to `out_placements`, `in_placements` should keep
+            a 1:1 mapping to the flattened input of `func`. If a redistribution is
+            required according to `in_placements` and `redistribute_inputs` is `False`,
+            an exception will be raised.
+        device_mesh (:class:`DeviceMesh`, optional):
+            the device mesh that all the :class:`DTensor`s are placed on. If not
+            specified, this will be inferred from the input :class:`DTensor`s' device
+            mesh. `local_map` requires every :class:`DTensor`s to be placed on the same
+            device mesh.
+        redistribute_inputs (bool, optional):
+            the bool value indicating whether to reshard the input :class:`DTensor`s when
+            their placements are different from the required input placements. If this
+            value is `False` and some :class:`DTensor` input has a different placement,
+            an exception will be raised. Default: `False`.
+
+    Returns:
+        A `Callable` that applies `func` to each local shard of the input :class:`DTensor`
+        and returns a :class:`DTensor` constructed from the return value of `func`.
+
+    Raises:
+        AssertionError: If the input :class:`DTensor`s are not placed on the same device
+        mesh, or if they are placed on a different device mesh than the `device_mesh`
+        argument passed in.
+
+        AssertionError: If the placements in `out_placements` for non-:class:`DTensor`
```

Comment:
I don't quite get this, could you rephrase this a bit to make it easier to understand?